### PR TITLE
Editor flow changes, dragging + `inverted` calculation

### PR DESCRIPTION
**Pull request by kaglowka**
- ported annotator's `move` to new `src`
- removed `EditorContainer`
- changed `Editor` data flow
   - `Editor` receives `form` in props; The filling of form (an empty one or values from an existing annotation) happens in the `connect` layer. 
- restored `Editor` dragging using an intermediate `DraggableWidget` component that accepts refs to the draggable part of the editor.
- `inverted` calculation (unfortunately more convoluted than you could wish)
    - `inverted` calculation takes place after the first widget render. For a split second, the component is `hidden` but rendered, which is used to check if it fits the window size. It is then rendered visible with `inverted` state set. 
    - Widget / Editor `inverted` properties are now calculated in components' local state and are invisible in the Redux global state

### CODE REVIEW

File: src/components/editor/Editor.tsx [hunk 1]
```diff
@@ -67,22 +99,27 @@ class Editor extends React.Component<Partial<IEditorProps>, Partial<IEditorState
...
     };
   }
 
+  moverElement: RefObject<HTMLDivElement>;
+  dragTracker: DragTracker;
+
   constructor(props: IEditorProps) {
     super(props);
+    console.log(props);
     this.state = Editor.stateFromProps(props);
```
Comment (lgodziejewski) [inline src/components/editor/Editor.tsx]: if the function was called `getDerivedStateFromProps`, then you wouldn't have to invoke it manually as it is part of the component lifecycle

https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

File: src/components/editor/Editor.tsx [hunk 2]
```diff
@@ -1,63 +1,95 @@
...
 
+  console.log(selectEditorState(state));
+
   return {
-    invertedX,
-    invertedY,
-    locationX,
-    locationY,
-    visible,
+    editor: {
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: That's true. I just started tweaking with the editor and learned that it pays off to keep your props shallow :) I'm only just improving on this.

File: src/components/editor/Editor.tsx [hunk 3]
```diff
@@ -1,63 +1,95 @@
...
-import Widget from 'components/widget';
-
-import { AnnotationPriorities, annotationPrioritiesLabels } from '../consts';
-import PriorityButton from './priority-button/PriorityButton';
+import {AnnotationPriorities, annotationPrioritiesLabels} from '../consts';
 import styles from './Editor.scss';
+import {DragTracker, IVec2} from "../../utils/move";
+import PriorityButton from "./priority-button/PriorityButton";
+import {Modal, Popup} from "semantic-ui-react";
+import Widget from "../widget/Widget";
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: Now I see where it came from. When I was solving (many) conflicts, I sometimes re-added the imports which was quicker than unifying my and your imports. I didn't think of the consequences for the commit history, though.

File: src/store/widgets/reducers.ts [hunk 4]
```diff
@@ -20,10 +20,6 @@ const initialWidgetState = {
     x: 0,
     y: 0,
   },
-  inverted: {
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/store/widgets/reducers.ts]: :)

File: src/store/annotations/reducers.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
+const initialState = [];
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/store/annotations/reducers.ts]: Sounds reasonable, but what would be the path to the actual annotation list? (assuming we wouldn't want `state.annotations.annotations` ;) )

File: src/containers/App.tsx [hunk 6]
```diff
@@ -1,10 +1,8 @@
...
-
-import AnnotationViewModel from 'models/AnnotationViewModel';
-
-import Editor from 'components/editor';
-import Menu from 'components/menu';
-import Viewer from 'components/viewer';
+import Menu from "../components/menu/Menu";
+import Viewer from "../components/viewer/Viewer";
+import AnnotationViewModel from "../models/AnnotationViewModel";
+import Editor from "../components/editor/Editor";
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/containers/App.tsx]: Yes, they were. Sorry, I'll work on the editor settings. I'll rework the imports for this very PR.

File: src/components/editor/Editor.tsx [hunk 7]
```diff
@@ -1,63 +1,95 @@
...
+    priority: model.priority || AnnotationPriorities.NORMAL,
+    comment: model.comment || '',
+    referenceLink: model.referenceLink || '',
+    referenceLinkTitle: model.referenceLinkTitle || '',
+  };
+}
+
 @connect((state) => {
+  let form;
+  if (state.annotationId) {
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: Sorry, meant `state.widgets.editor.annotationId` (I'll include it in the next PR).
That's what happens when you write an `if` you cannot test.

File: src/components/editor/Editor.tsx (hunk 1 above)
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: Done

File: src/components/editor/Editor.tsx [hunk 8]
```diff
@@ -99,6 +136,53 @@ class Editor extends React.Component<Partial<IEditorProps>, Partial<IEditorState
...
+  onDrag = (delta: IVec2) => {
+    this.setState({
+      locationX: this.state.locationX + delta.x,
+      locationY: this.state.locationY + delta.y,
+    });
+    return true;
+  }
+
+  onMouseUp = () => {
+    this.setState({isDragged: false});
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: No. Removed it.

File: src/components/widget/Widget.tsx [hunk 9]
```diff
@@ -1,50 +1,100 @@
...
+  invertedX: boolean;
+  invertedY: boolean;
+  calculateInverted: boolean;
+}
 
-  static classes = {
-    invertedX: 'inverted-x',
-    invertedY: 'inverted-y',
-  };
+export default class Widget extends React.Component<Partial<IWidgetProps>,
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/widget/Widget.tsx]: During dragging, we don't want the window to flip, so `Widget` needs some kind of contextual information, if it should be ready to recalculate inversion or not. I removed `calculateInverted` it from `Editor` but decided to retain this minimum of control over `Widget` for now.

File: src/components/editor/Editor.tsx [hunk 10]
```diff
@@ -1,63 +1,95 @@
...
   invertedY: boolean;
   locationX: number;
   locationY: number;
-  annotation: AnnotationViewModel;
+  /*
+   * calculateInverted - (overwrites invertedX and invertedY)
+   * if true, the widget horizontal or vertical inversion will be calculated based on the window location
+   * after the component is rendered for the first time after prop change
+   */
+  calculateInverted: boolean;
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: I removed it from `Editor` component. (but left it on `Widget`)

File: src/components/editor/Editor.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/components/editor/Editor.tsx]: sorry, meant `state.wieditor.annotationId`. That's what happens when you write an `if` you can't test.

**Pull request merged**